File: testbed/nodecpp/calculate/results.js
const addon = require("./cpp/build/Release/results");

var location = {
    country : "Germany",
    samples : [
        {month : "January",   results : 1000},
        {month : "February",  results : 2000},
        {month : "March",     results : 3000},
        {month : "April",     results : 4000},
        {month:  "May",       results : 5000}
    ]
};

// //var random = () => { return Math.round(1000000 * Math.random() * 100) / 100;}

// var random = function(){
//     return Math.round(1000000 * Math.random() * 100) / 100;
// }().toFixed(2);
// //console.log(random());

var countries = ["Germany", "USA", "France", "Italy", "Spain", "Canada", "Australia", "England", "Japan", "South Africa"];
var locations = [];
for (var i = 0; i < 10; i++){
    var loc = {
        country : countries[i],
        samples : [
            {month : "January",  results : 2000},
            {month:  "February", results: 3000},
            {month:  "March",    results: 4000},
            {month:  "April",    result: 5000}
        ]
    }
    locations.push(loc);
}

var results = addon.calculate_results(locations);
console.log(results);
var i = 0;
results.forEach((result) => {
            console.log("Result for Location " + i);
            console.log("---------------------------");
            console.log("\tCountry: " + locations[i].country);
            console.log("\tMean Sales: " + result.mean);
            console.log("\tMedian Sales: " + result.median);
            console.log("\tStandard Dev.: " + result.standard_deviation);
            console.log("\tNumber Samples: " + result.n);
            console.log();
            i++;
});




// const results = addon.avg_sales(location);
// console.log(`Average Results = ${results} €`);

// const data = addon.data_country(location);

// console.log("Mean = " + data.mean);
// console.log("Median = " + data.median);
// console.log("Standard Deviation = " + data.standard_deviation);
// console.log("Count = " + data.n);

File: testbed/nodecpp/calculate/cpp/results.cpp
#include "results.h"
#include <algorithm>
#include <cmath>

bool operator <(const sample &s1, const sample &s2){
    return s1.results < s2.results;
}

double avg_sales(location &loc){
    double total = 0;
    for (const auto &sample : loc.samples){
        total += sample.results;       
    }
    return total / loc.samples.size();
}

country_result calc_country_stats(location &loc){
        country_result result;
        float ss = 0;
        double total = 0;
        result.n = loc.samples.size();

        // for (const auto &sample : loc.samples){
        //     total += sample.results;
        // }
        for (int i = 0; i < result.n - 1; i++){
                total += loc.samples[i].results;
        }
        //total = loc.samples[0].results + loc.samples[1].results + loc.samples[2].results;// + loc.samples[3].results;
        result.mean = loc.samples[3].results; //total / loc.samples.size();

        for (const auto &sample : loc.samples){
            ss += pow(sample.results - result.mean, 2);
        }
        result.standard_deviation = sqrt(ss/result.n - 1);

        std::sort(loc.samples.begin(), loc.samples.end());

        if (result.n % 2 == 0){
            result.median = (loc.samples[result.n/2 -1].results + loc.samples[result.n / 2].results) / 2;
        }
        else {
            result.median = loc.samples[result.n / 2].results;
        }
        return result;
}


